alpha/alpha-core/src/main/java/org/apache/servicecomb/saga/alpha/core/TxConsistentService.java

```diff
@@ -44,7 +58,32 @@ public void handle(TxEvent event) {
   }
 
   private void compensate(TxEvent event) {
-    List<TxEvent> events = eventRepository.findStartedTransactions(event.globalTxId(), EventType.TxStartedEvent.name());
+    List<TxEvent> events = eventRepository.findStartedTransactions(event.globalTxId(), TxStartedEvent.name());
     events.forEach(omegaCallback::compensate);
+    eventsToCompensate.computeIfAbsent(event.globalTxId(), (v) -> {
+      CopyOnWriteArraySet<String> eventSet = new CopyOnWriteArraySet<>();
+      events.forEach(e -> eventSet.add(getUniqueEventId(e)));
```

Comment:
what's the performance characteristics of CopyOnWrite?

Reply:
There is no need to use CopyOnWriteArraySet as the set is only write once. Already used HashSet instead.